### PR TITLE
Reduce stack depth in StateT

### PR DESCRIPTION
**Pull request by johnynek**
I hit some stack overflows using traverse with StateT. I don't think this will always solve the issue, but I tried to minimize stack depth in a number of cases.

### CODE REVIEW

**Comment by johnynek** [comment]
looks like a flakey test (Futures), restarting.

File: core/src/main/scala/cats/data/StateT.scala [hunk 1]
```diff
@@ -70,9 +77,12 @@ final class StateT[F[_], S, A](val runF: F[S => F[(S, A)]]) extends Serializable
    * Like [[map]], but also allows the state (`S`) value to be modified.
    */
   def transform[B](f: (S, A) => (S, B))(implicit F: Monad[F]): StateT[F, S, B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/StateT.scala]: this only needs `Functor[F]` in this formulation.

File: core/src/main/scala/cats/data/StateT.scala [hunk 2]
```diff
@@ -12,25 +12,32 @@ import cats.syntax.either._
 final class StateT[F[_], S, A](val runF: F[S => F[(S, A)]]) extends Serializable {
 
   def flatMap[B](fas: A => StateT[F, S, B])(implicit F: Monad[F]): StateT[F, S, B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/StateT.scala]: this only needs `FlatMap[F]` in this formulation, but I didn't want to change binary compatibility for a negligible gain.

File: core/src/main/scala/cats/data/StateT.scala [hunk 3]
```diff
@@ -12,25 +12,32 @@ import cats.syntax.either._
...
+    StateT.applyF(F.map(runF) { sfsa =>
+      sfsa.andThen { fsa =>
+        F.flatMap(fsa) { case (s, a) =>
           fas(a).run(s)
         }
-      })
+      }
+    })
 
   def flatMapF[B](faf: A => F[B])(implicit F: Monad[F]): StateT[F, S, B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/StateT.scala]: only needs `FlatMap[F]` in the current formulation.

File: core/src/main/scala/cats/data/StateT.scala [hunk 4]
```diff
@@ -12,25 +12,32 @@ import cats.syntax.either._
...
-          F.map(faf(a))((s, _))
-        }
+    StateT.applyF(F.map(runF) { sfsa =>
+      sfsa.andThen { fsa =>
+        F.flatMap(fsa) { case (s, a) => F.map(faf(a))((s, _)) }
       }
-    )
+    })
 
   def map[B](f: A => B)(implicit F: Monad[F]): StateT[F, S, B] =
```
Comment (johnynek) [inline core/src/main/scala/cats/data/StateT.scala]: if we fix `transform` to only use `Functor[F]` then this only needs `Functor[F]`, which is fitting for `map`.

File: core/src/main/scala/cats/data/StateT.scala (hunk 2 above)
Comment (non) [inline core/src/main/scala/cats/data/StateT.scala]: I think that makes sense for now. Maybe add a comment to that effect? If we ever want to tighten things up later that will be handy.

**Comment by non** [comment]
These changes seem positive to me.

I think we should either document all the places where type class bounds could be tightened, or just do the change now. While breaking compatibility is a drag, better now than at 1.0.

👍 whichever way you go.

**Comment by johnynek** [comment]
Okay, I lowered everything about as much as I think you can. Nice to only need `Functor[F]` in many places, which makes also worthwhile to add a `Functor[StateT[F, S, ?]]` instance.

**Comment by non** [comment]
👍

**Comment by johnynek** [comment]
@adelbertc what do you think?

**Comment by adelbertc** [comment]
LGTM 👍

**Pull request merged**